Github action for testing latest quarkus `master` with latest graal `master`

zakkak opened the pull request:
This PR adds a github action for building the latest graal and then using it to build and test the latest quarkus.

The goal of this github action is to assist the graalvm team in detecting changes that break quarkus.

zakkak commented:
This github action will be ultimately set to run nightly. It is currently set to run on push for testing purposes till it gets reviewed.

Comments on how to improve it are welcome :)

jerboaa commented:
> We use the [Oracle Contributor Agreement](http://www.oracle.com/technetwork/community/oca-486395.html) to make the copyright of contributions clear. We don't have a record of you having signed this yet, based on your email address fzakkak -(at)- redhat -(dot)- com. You can sign it at that link.
> 
> If you think you've already signed it, please comment below and we'll check.

The @redhat.com email address was the clue. He should be covered by the Red Hat OCA as he works for Red Hat.

ansalond commented:
Hi @zakkak, thanks for the PR. I'll add a couple of comments.

maxandersen commented on .github/workflows/quarkus.yml:
```diff
@@ -0,0 +1,335 @@
+
+jobs:
+  build-quarkus-and-graalvm:
+    name: Nightly quarkus and GraalVM build
```
when done on push this isn't really nightly ;)

adinn replied:
I believe Foivos commented on that earlier: "This github action will be ultimately set to run nightly. It is currently set to run on push for testing purposes till it gets reviewed."

maxandersen replied:
sorry I missed that.

Reason for the nightly is that it takes too long to verify ? I would assume run on pull-requests would be useful or ?

zakkak replied:
I just assumed we wanted it Nightly because our initial discussion was about nightly builds.
Running on pull-requests would certainly prevent breaking changes from making it to `master`. 

The whole workflow needs something less than 1h and 30m at the moment.

maxandersen replied:
sorry if I made it sound initially as a nightly build - that was not my intention. 

Intention was to have *something* on graalvm side that can be used as early warning signs if some change in graalvm master affects quarkus (and vice versa) - preferably that would be on `push` but if deemed too long/too flakey nightly builds would do.

My preference is to have it at least on pull-requests coming in and on push to master (or nightly) would at least catch those things that are not coming in via pull-requests too.

ansalond replied:
I propose to start with a nightly task that tests the tip of `graal` with the latest stable release of Quarkus.
This would make our tests more stable and reproducible.
If we see that this is not enough, we can easily change policy.

zakkak replied:
https://github.com/oracle/graal/commit/55990119a23baeadf3967ae3565aea82069d5352 makes it build using the latest github release of quarkus instead of the tip (this includes candidate releases, but I think this is OK, please let me know if you disagree).

I am letting it run on push till we _finalize_ this PR.

ansalond replied:
Thanks!

zakkak replied:
@ansalond Just a reminder that before merging I need to resolve the above ^^

zakkak replied:
@ansalond I have now configured it to run nightly at 3am and when changes affecting the workflow are pushed.
Note that this workflow tests latest graal with the latest quarkus release from https://github.com/quarkusio/quarkus/releases (no longer testing pre-releases)

zakkak commented on .github/workflows/quarkus.yml:
```diff
@@ -0,0 +1,335 @@
+env:
+  # Workaround testsuite locale issue
+  LANG: en_US.UTF-8
+  JVMCI_VERSION: jvmci-20.1-b01
+  JDK_VERSION: 11.0.6
+  JDK_VERSION_SUFFIX: "+9"
```
Yes that sounds like a good idea :)

I guess the above link points to Oracle's internal repository and is the same as:
https://github.com/oracle/graal/blob/master/common.json#L8

So I will set it up to parse the common.json from the checked out repository.

ansalond replied:
Yes, sorry, you are right.

zakkak replied:
Fixed in https://github.com/oracle/graal/pull/2349/commits/fa2e0a9e1caaced8465cc35c296ddcc88fe9338d

zakkak commented on .github/workflows/quarkus.yml:
```diff
@@ -0,0 +1,335 @@
+        repository: quarkusio/quarkus.git
+        ref: master
+        path: quarkus
+    # actions/setup-java setups a zulu community distibution of OpenJDK
+    # apt-get install openjdk-11-jdk installs the same zulu distribuion on github runners
+    # so we use joschi/setup-jdk instead with an sha for added security since tags can be updated
+    - uses: joschi/setup-jdk@9dc354d4a8f5781ba8880dc9e6a7bdcc7845e3e7
+      with:
+        java-version: 11
```
AFAIK we prefer using the latest OpenJDK release for anything that is not directly related to GraalVM (e.g. native images and polyglot support). 

@adinn @maxandersen your thoughts?

maxandersen replied:
@zakkak You are right from a Quarkus perspective we would be validating Quarkus on top of OpenJDK, but from GraalVM project it would make sense they test Quarkus on a labsjdk/graalvm - ultimately we would want both test runs but to start I would let graalvm team pick their priority.

just be warned we (Quarkus) doesn't target labsjdk; but happy to get PR's if for someone reason Quarkus does something bad that affects it :)

zakkak replied:
Resolved in https://github.com/oracle/graal/commit/c07a30553765e65d5b00acc691493842666f8d2c

maxandersen commented on .github/workflows/quarkus.yml:
```diff
@@ -0,0 +1,335 @@
+        path: maven-repo.tgz
+
+  native-tests:
+    name: Native Tests - ${{matrix.category}}
```
worth noting that these matrix tests are a current snapshot from quarkus own native tests and they will change over time - we'll want to try and find a more stable way of defining the tests but for now this should do.

galderz replied:
+100

How would be achieve this though? Seems to me Quarkus should list/define these groups. However, don't know how Quarkus could maintain that list in a way that it doesn't break the moment a new integration test is added or removed.

zakkak replied:
Started discussion on [the quarkus-dev mailing list](https://groups.google.com/forum/#!topic/quarkus-dev/_KZe1ydg4C0).

I suggest we proceed with merging this PR and when we find a better way of doing this we can create a new PR.

zakkak commented:
I have also tried building graalvm with `--exclude-components=nju,svml,polynative,lg,llp,poly,tflm,mjdksl`
(https://github.com/zakkak/graal/commit/ea254b9d732f4468dc6b08bcb8af5929c69516ac) and the result works fine for quarkus. Would you like me to integrate this in this workflow or do you prefer to test with the _defaults_ (i.e. `mx build`)?

ansalond commented:
> Would you like me to integrate this in this workflow or do you prefer to test with the _defaults_ (i.e. `mx build`)?

I'm fine with running `mx build` from `graal/substratevm`. If you prefer to reduce the set of built components, I would specify an inclusion list based on component names rather than an exclusion list based on component short names. These things change often, and the inclusion list should be more stable.

Looking at your list of excluded components, I'd run:
```
mx --components="Native Image Configure Tool" build
```
Which includes:
```
$ mx --components="Native Image Configure Tool" graalvm-show
GraalVM distribution: GRAALVM_8BEC66804A_JAVA8
Version: 20.2.0-dev
Config name: None
Components:
 - Native Image Configure Tool ('nic', /svm)
 - Native Image ('ni', /svm)
 - SubstrateVM ('svm', /svm)
 - Native Image licence files ('nil', /svm)
 - GraalVM compiler ('cmp', /graal)
 - Truffle Macro ('tflm', /truffle)
 - Truffle NFI ('nfi', /nfi)
 - Truffle ('tfl', /truffle)
 - Graal SDK ('sdk', /graalvm)
Launchers:
 - native-image-configure (bash)
 - native-image (bash)
Libraries:
 - libnative-image-agent.so (skipped)
No installable
No standalone
```

However, this would be closer to what we ship (no `native-image-configure` and a native image for `native-image`):
```
$ mx --components="Native Image" --native-images=native-image build
```
```
$ mx --components="Native Image" --native-images=native-image graalvm-show
GraalVM distribution: GRAALVM_3E9B09AF6B_JAVA8
Version: 20.2.0-dev
Config name: None
Components:
 - Native Image ('ni', /svm)
 - SubstrateVM ('svm', /svm)
 - Native Image licence files ('nil', /svm)
 - GraalVM compiler ('cmp', /graal)
 - Truffle Macro ('tflm', /truffle)
 - Truffle NFI ('nfi', /nfi)
 - Truffle ('tfl', /truffle)
 - Graal SDK ('sdk', /graalvm)
Launchers:
 - native-image (native)
Libraries:
 - libnative-image-agent.so (skipped)
No installable
No standalone
```

ansalond commented:
I'm waiting for the confirmation that we can add GitHub actions to this repository. I'll keep you updated.

hanza-23 commented:
Good news: we can add GitHub actions to this repository. I'll work on integrating this PR.

ansalond commented:
Good news: we can add GitHub actions to this repository. I'll work on integrating this PR.

zakkak commented:
Hi @ansalond I have updated the matrix, squashed the changes in a single commit and rebased on the latest `master`, let's see how it goes before merging (you can see it running [here](https://github.com/zakkak/mandrel/actions/runs/211906644) )

ansalond commented on .github/workflows/quarkus.yml:
```diff
@@ -0,0 +1,371 @@
+        with:
+          java-version: 11
+      - name: Reclaim Disk Space
+        run: .github/ci-prerequisites.sh
```
@zakkak, please add `.github/ci-prerequisites.sh` or remove this step.

```
Run .github/ci-prerequisites.sh
/home/runner/work/_temp/cfae83e8-7803-4c08-9ef8-59de516c2f55.sh: line 1: .github/ci-prerequisites.sh: No such file or directory
##[error]Process completed with exit code 127.
```
https://github.com/zakkak/mandrel/runs/993279323?check_suite_focus=true

zakkak commented:
@ansalond please check latest run at https://github.com/zakkak/mandrel/runs/994834795?check_suite_focus=true

ansalond commented:
Thanks! I see some failures, should we open dedicated issues?

zakkak commented:
Yes, I think the right way to move forward is to open dedicated issues where the graal and the quarkus communities can work together.
Could we also have a `quarkus` label or similar for those issues to make it easy to track?

ansalond commented:
Perfect! I created the `quarkus` label, please open the required issues. Let me know when you want to merge this PR. We just need to change the conditions that trigger the action.

zakkak commented:
Great, thanks @ansalond !

I think the PR is ready to be merged. The conditions that trigger the action are already changed.
As mentioned in https://github.com/oracle/graal/pull/2349#discussion_r471766309 the action now runs nightly at 3am, and when changes to the workflow are pushed or a PR is created.
See https://github.com/oracle/graal/pull/2349/files#diff-dacc38c76a716b588c8615e697332287R3-R11
